### PR TITLE
P3 skeleton framework

### PR DESCRIPTION
**Pull request by rorlija1**
This PR aims to add compatibility in KiD to test the 1D framework described in case 1 of [Cholette et al 2019](https://doi.org/10.1175/JAS-D-18-0278.1) using P3.

Only P3 terminal velocity is included with this PR.

For future reference, the configuration I've been using for testing is as follows:
```
config = parse_commandline()
config["precipitation_choice"] = "PrecipitationP3"
config["moisture_choice"] = "MoistureP3"
config["n_elem"] = 24
config["z_max"] = 3000
config["t_end"] = 75
config["w1"] = 0
config["dt"] = Float64(0.5)
```

Note: the `p3_boundary_condition`  command line argument controls whether we use an initial signal of ice or we introduce ice mass via the boundary, and the same argument is also used to customize the characteristics of the initial/boundary condition particles.

### CODE REVIEW

**Comment by rorlija1** [comment]
@sajjadazimi I've just tagged you for review on this PR—I haven't added tests for P3 and have not squashed/rebased yet, but I'm hoping to get your feedback on the code as I prepare to merge this. Thanks!!

File: src/Common/tendency.jl [hunk 1]
```diff
@@ -62,7 +89,7 @@ end
     @. ts = TD.PhaseNonEquil_ρTq(thermo_params, ρ, T, PP(q_tot, q_liq, q_ice))
     @. p = TD.air_pressure(thermo_params, ts)
     @. θ_dry = TD.dry_pottemp(thermo_params, T, ρ_dry)
-    @. θ_liq_ice = TD.liquid_ice_pottemp(thermo_params, ts)
+    @. θ_liq_ice = θ_dry #TD.liquid_ice_pottemp(thermo_params, ts) - prevents error...
```
Comment (sajjadazimi) [inline src/Common/tendency.jl:92]: Why did you change this line?

File: src/K1DModel/tendency.jl [hunk 2]
```diff
@@ -269,6 +269,25 @@ end
...
+    # advect aerosols
+
+    If = CC.Operators.InterpolateC2F()
+    ∂ = CC.Operators.DivergenceF2C(
+        bottom = CC.Operators.SetValue(CC.Geometry.WVector(aux.prescribed_velocity.ρw0 * aux.q_surf)),
+        top = CC.Operators.Extrapolate(),
+    )
+    @. dY.N_aer += -∂(aux.prescribed_velocity.ρw / If(aux.thermo_variables.ρ) * If(Y.N_aer))
+
+    if Bool(aux.kid_params.qtot_flux_correction)
```
Comment (sajjadazimi) [inline src/K1DModel/tendency.jl:284]: remove this condition. Always apply fcc here

File: src/K1DModel/tendency.jl [hunk 3]
```diff
@@ -393,3 +412,179 @@ end
...
+    @. ρq_liqonice = Y.ρq_liqonice
+    @. N_ice = Y.N_ice
+    @. B_rim = Y.B_rim
+
+    # choose flux characteristics
+    _q_flux = FT(0.65e-4)
+    _N_flux = FT(40000)
+    _F_rim = FT(0.2)
+    _F_liq = FT(0.2)
+    _ρ_r_init = FT(900)
```
Comment (sajjadazimi) [inline src/K1DModel/tendency.jl:440]: These should not be hardcoded

File: src/K1DModel/tendency.jl [hunk 4]
```diff
@@ -393,3 +412,179 @@ end
...
+
+    _flux = FT(0.5)
```
Comment (sajjadazimi) [inline src/K1DModel/tendency.jl:442]: same here

File: test/experiments/KiD_driver/run_KiD_simulation.jl [hunk 5]
```diff
@@ -104,6 +104,33 @@ function run_KiD_simulation(::Type{FT}, opts) where {FT}
...
+        # config["precipitation_choice"] = "PrecipitationP3"
+        # config["moisture_choice"] = "MoistureP3"
+        # config["n_elem"] = 24
+        # config["z_max"] = 3000
+        # config["t_end"] = 75
+        # config["w1"] = 0
+        # config["rv_0"] = 0
+        # config["rv_1"] = 0
+        # config["rv_2"] = 0
+        # config["dt"] = Float64(0.5)
```
Comment (sajjadazimi) [inline test/experiments/KiD_driver/run_KiD_simulation.jl:118]: These lines should be removed before merging

**Review by sajjadazimi — COMMENTED** [review]
Thank you!
It looks good to me. I left some small comments.

File: src/K1DModel/tendency.jl (hunk 3 above)
Comment (rorlija1) [reply src/K1DModel/tendency.jl:440]: Right—thanks. Would it be acceptable to add them as elements of the PrecipitationP3 struct that can be initialized in run_KiD_simulation? If so, would the best place be to add them as command line arguments in parse_commandline? Otherwise, I could add them in aux as command line arguments as well. Maybe a command line argument that is used in aux could be the cleanest way to do it?

File: src/Common/initial_condition.jl [hunk 6]
```diff
@@ -281,3 +281,100 @@ function cloudy_initial_condition(pdists, ip, k = 1)
...
+    q_rim::FT = F_rim * (q_ice - q_liqonice)
+
+    # 2M warm rain state variables
+    q_liq::FT = FT(0)
+    N_liq::FT = FT(0)
+    q_rai::FT = FT(0)
+    N_rai::FT = FT(0)
+
+    # q_tot - single p3 category + 2M categories
+    q_tot::FT = q_ice + q_liq + q_rai
```
Comment (trontrytel) [inline src/Common/initial_condition.jl:309]: Maybe we could add some water vapor to this setup. Such that q_tot = q_vap + q_liq + q_ice?

File: src/Common/tendency.jl (hunk 1 above)
Comment (trontrytel) [reply src/Common/tendency.jl:92]: I wonder if the fact that we are running it with q_vap = 0 is the problem

File: src/K1DModel/tendency.jl (hunk 3 above)
Comment (trontrytel) [reply src/K1DModel/tendency.jl:440]: I think they would fit better in the case setup. Since they only come about because the case specifies the top fluxes. In an actual simulation we would make sure that our domain is high enough that we don't have to worry about anything else falling in

File: test/experiments/KiD_driver/run_KiD_simulation.jl [hunk 7]
```diff
@@ -104,6 +104,33 @@ function run_KiD_simulation(::Type{FT}, opts) where {FT}
...
+        cloudy_params = nothing
+        init = map(
+            coord -> CO.p3_initial_condition(
```
Comment (trontrytel) [inline test/experiments/KiD_driver/run_KiD_simulation.jl:121]: why do we need a special case for the P3 scheme? Could it be handled in the same way as other bulk microphysics cases?

**Comment by trontrytel** [comment]
Could we also add this case to our buildkite jobs: https://github.com/CliMA/KinematicDriver.jl/blob/main/.buildkite/pipeline.yml

File: src/Common/initial_condition.jl (hunk 6 above)
Comment (rorlija1) [reply src/Common/initial_condition.jl:309]: Sure! Would you say I should add some initial q_vap using the same cosine initial condition as the ice_bump (line 294) uses? If so, is there a reasonable amplitude I should use?
I'm guessing water vapor will be used in ice nucleation but that otherwise it will hang out where it starts (as long as we don't prescribe an updraft velocity)

File: src/K1DModel/tendency.jl (hunk 3 above)
Comment (rorlija1) [reply src/K1DModel/tendency.jl:440]: Ah, I see your reasoning: just to make sure, do you mean this [function](https://github.com/CliMA/KinematicDriver.jl/pull/185/files#diff-e11ad7cffaff946aac755470d2db05dbc518553d52a499dd9134587cb701e70bR285)? Maybe I can add the flux characteristics hardcoded into the initial condition, and then pass them to aux this way so that they can be used in the advection tendency? If I do this, then I can make it so that the particles are either initialized with ice_bump in the initial condition or introduced through the boundary. Or I can add them as command line arguments that are used by the initial condition. Do either of you have a preference as to which might be better?

File: test/experiments/KiD_driver/run_KiD_simulation.jl (hunk 7 above)
Comment (rorlija1) [reply test/experiments/KiD_driver/run_KiD_simulation.jl:121]: I think if we want to initialize particles with specific P3 properties, we have to have another initial condition function, and in that initial condition, we specify our own temperature and pressure gradients, and we can add vapor pressure, etc. In the future we could try other initial conditions, but maybe for now this way is easiest for testing if you agree?

File: src/Common/tendency.jl (hunk 1 above)
Comment (rorlija1) [reply src/Common/tendency.jl:92]: Just changed it back to using θ_liq_ice from Thermodynamics, and no error! So whatever was causing it when I originally commented it out must be fixed now

File: src/Common/initial_condition.jl (hunk 6 above)
Comment (rorlija1) [reply src/Common/initial_condition.jl:309]: Ah—update: I could also use the same profile of water vapor which I just noticed that's used in the normal 1D testing setup. I'm guessing that would be the best way forward.
